transformations/transformations/scc_revector.py

```diff
@@ -0,0 +1,98 @@
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+from loki.expression import symbols as sym
+from loki import Transformation, FindNodes, ir, NestedTransformer, Transformer
+from transformations.scc_base import SCCBaseTransformation
+
+__all__ = ['SCCRevectorTransformation']
+
+class SCCRevectorTransformation(Transformation):
+    """
+    A transformation to wrap thread-parallel IR sections within a horizontal loop.
+    This transformation relies on markers placed by :any:`SCCDevectorTransformation`.
+
+    Parameters
+    ----------
+    horizontal : :any:`Dimension`
+        :any:`Dimension` object describing the variable conventions used in code
+        to define the horizontal data dimension and iteration space.
+    """
+
+    def __init__(self, horizontal):
+        self.horizontal = horizontal
+        self._processed = {}
+
+    @classmethod
+    def wrap_vector_section(cls, section, routine, horizontal):
+        """
+        Wrap a section of nodes in a vector-level loop across the horizontal.
+
+        Parameters
+        ----------
+        section : tuple of :any:`Node`
+            A section of nodes to be wrapped in a vector-level loop
+        routine : :any:`Subroutine`
+            The subroutine in the vector loops should be removed.
+        horizontal: :any:`Dimension`
+            The dimension specifying the horizontal vector dimension
+        """
+
+        # Create a single loop around the horizontal from a given body
+        v_start = routine.variable_map[horizontal.bounds[0]]
+        v_end = routine.variable_map[horizontal.bounds[1]]
+        index = SCCBaseTransformation.get_integer_variable(routine, horizontal.index)
+        bounds = sym.LoopRange((v_start, v_end))
+
+        # Ensure we clone all body nodes, to avoid recursion issues
+        vector_loop = ir.Loop(variable=index, bounds=bounds, body=Transformer().visit(section))
+
+        # Add a comment before the pragma-annotated loop to ensure
+        # we do not overlap with neighbouring pragmas
+        return (ir.Comment(''), vector_loop)
+
+    def transform_subroutine(self, routine, **kwargs):
+        """
+        Apply SCCRevector utilities to a :any:`Subroutine`.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            Subroutine to apply this transformation to.
+        role : string
+            Role of the subroutine in the call tree; should be ``"kernel"``
+        """
+
+        # TODO: we only need this here until the scheduler can combine multiple transformations into single pass
+        # Bail if routine has already been processed
+        if self._processed.get(routine, None):
```

Comment:
Same here

Reply:
Yup, will replace all of these with the existing mechanism. Thanks again!